qmi/instruments/thorlabs/apt_protocol.py

```diff
@@ -0,0 +1,249 @@
+    P_MOT_SB_INSTRERROR = 0x00008000  # unable to execute command
+    P_MOT_SB_INTERLOCK = 0x00010000  # used in controllers where a seperate signal is used to enable the motor
+    P_MOT_SB_OVERTEMP = 0x00020000  # motor or motor power driver electronics reached maximum temperature
+    P_MOT_SB_BUSVOLTFAULT = 0x00040000  # low supply voltage
+    P_MOT_SB_COMMUTATIONERROR = 0x00080000  # problem with motor commutation. Can only be recovered with power cycle
+    P_MOT_SB_DIGIP1 = 0x00100000  # state of digital input 1
+    P_MOT_SB_DIGIP2 = 0x00200000  # state of digital input 2
+    P_MOT_SB_DIGIP4 = 0x00400000  # state of digital input 3
+    P_MOT_SB_DIGIP8 = 0x00800000  # state of digital input 4
+    P_MOT_SB_OVERLOAD = 0x01000000  # some form of motor overload
+    P_MOT_SB_ENCODERFAULT = 0x02000000  # encoder fault
+    P_MOT_SB_OVERCURRENT = 0x04000000  # motor exceeded continuous current limit
+    P_MOT_SB_BUSCURRENTFAULT = 0x08000000  # excessive current being drawn from motor power supply
+    P_MOT_SB_POWEROK = 0x10000000  # controller power supplies operating normally
+    P_MOT_SB_ACTIVE = 0x20000000  # controller executing motion commend
+    P_MOT_SB_ERROR = 0x40000000  # indicates an error condition
+    P_MOT_SB_ENABLED = 0x80000000  # motor output enabled, with controller maintaining position
+
+
+class AptMessageId(Enum):
+    """Message IDs for devices using the APT protocol."""
+
+    HW_REQ_INFO = 0x0005
+    HW_GET_INFO = 0x0006
+    MOD_IDENTIFY = 0x0223
+    MOD_SET_CHANENABLESTATE = 0x0210
+    MOD_REQ_CHANENABLESTATE = 0x0211
+    MOD_GET_CHANENABLESTATE = 0x0212
+    HW_START_UPDATEMSGS = 0x0011
+    HW_STOP_UPDATEMSGS = 0x0012
+    MOT_MOVE_HOME = 0x0443
+    MOT_MOVE_HOMED = 0x0444
+    MOT_MOVE_ABSOLUTE = 0x0453
+    MOT_MOVE_COMPLETED = 0x0464
+    MOT_SET_EEPROMPARAMS = 0x04B9
+    MOT_REQ_USTATUSUPDATE = 0x0490
+    MOT_GET_USTATUSUPDATE = 0x0491
+    MOT_MOVE_JOG = 0x046A
+    POL_SET_PARAMS = 0x0530
+    POL_REQ_PARAMS = 0x0531
+    POL_GET_PARAMS = 0x0532
+
+
+class AptChannelState(Enum):
+    """Channel state"""
+
+    ENABLE = 0x01
+    DISABLE = 0x02
+
+
+class AptChannelJogDirection(Enum):
+    """Jog direction"""
+
+    FORWARD = 0x01
+    BACKWARD = 0x02
+
+
+class AptMessageHeaderWithParams(LittleEndianStructure):
+    """
+    This is the version of the APT message header when no data packet follows a header.
+    """
+
+    _pack_ = True
+    _fields_: List[Tuple[str, type]] = [
+        ("message_id", c_uint16),
+        ("param1", c_uint8),
+        ("param2", c_uint8),
+        ("dest", c_uint8),
+        ("source", c_uint8),
+    ]
+
+
+class AptMessageHeaderForData(LittleEndianStructure):
+    """
+    This is the version of the APT message header when a data packet follows a header.
+    """
+
+    _pack_ = True
+    _fields_: List[Tuple[str, type]] = [
+        ("message_id", c_uint16),
+        ("date_length", c_uint16),
```

Comment:
I suppose you should refer here to message data length, not message date as in day,month,year...?

Reply:
indeed!